### PR TITLE
Replace `Players` collection

### PR DESCRIPTION
**Pull request by InusualZ**
The game uses a collection of slot to store the players of the current layer. In some packets the server must send a valid player index and this index must reflect the position in the slot collection across all players in the same layer. With our previous collection (plain `list`) this wasn't possible because if at some point someone was removed from the collection, all other player index position would become invalid.

Now we implement that sort of collection with a list of slot (default `None`). When someone is added we give it the first empty (`None`) slot and when someone is removed we empty-out (set to `None`) the slot. With this any other player index position stay the same throughout their connection in the layer.

### CODE REVIEW

File: fmp_server.py [hunk 1]
```diff
@@ -530,13 +530,11 @@ def recvReqCircleJoin(self, packet_id, data, seq):
             self.sendAnsCircleJoin(0, 0, seq)
             return
 
-        # TODO: When appending get the index of the slot
-        circle.players.append(self.session)
+        player_index = circle.players.add(self.session)
+        assert player_index != -1, "City is full"
```
Comment (sepalani) [inline fmp_server.py:534]: Shouldn't it be `Circle is full` rather than `City`?

**Review by sepalani — COMMENTED** [review]
LGTM code-wise. I'm going to test it.

**Comment by sepalani** [comment]
After some testing, some exceptions are raised and need to be addressed for some packets:
 - [AnsLayerDetailSearchData](https://github.com/sepalani/MH3SP/blob/f1eb7ee1c962cd02eed43cead9ef8bfe1b2214de/mh/pat.py#L2117), `Players object doesn't support indexing`
 - [AnsUserSearchData](https://github.com/sepalani/MH3SP/blob/f1eb7ee1c962cd02eed43cead9ef8bfe1b2214de/mh/pat.py#L1166), `Session object is not iterable`

Do the packets related to the search features need this slot index? If not, leaving the search result as a list of sessions should be fine if there are no overhead.

**Comment by InusualZ** [comment]
Ahh I forgot to test after rebasing onto `dev`. Let me take a look.

**Comment by InusualZ** [comment]
> Do the packets related to the search features need this slot index? If not, leaving the search result as a list of sessions should be fine if there are no overhead.

To be honest, I haven't look into that, that much but I don't think the use this slot index. Since that slot index is mostly used when you are inside the city/circle and since the searching functionality don't actually interact with those layer I don't think they would ever use them.

**Comment by InusualZ** [comment]
Oops forgot to attach screenshot taken from the tests

*User Search*
image

*City Search*
image

**Review by sepalani — APPROVED** [review]
LGTM

**Pull request merged**